odoo/sql_db.py

```diff
@@ -100,6 +101,64 @@ def check(f, self, *args, **kwargs):
         raise psycopg2.OperationalError('Unable to use a closed cursor.')
     return f(self, *args, **kwargs)
 
+class Savepoint:
+    """ Reifies an active breakpoint, allows :meth:`BaseModel.savepoint` users
+    to internally rollback the savepoint (as many times as they want) without
+    having to implement their own savepointing, or triggering exceptions.
+
+    While the savepoint name is created during (and accessible after)
+    initialisation, the actual savepoint is only defined in ``__enter__``, and
+    only valid for the context-managed span. The savepoint is automatically
+    released ("committed") on context exit, unless an exception was raised in
+    which case it's rolled back first.
+
+    It can also be covered by `closing`, with similar semantics to
+    :class:`Cursor`: closing a savepoint will unconditionally roll it back
+    before releasing it.
```

Comment:
This sentence suggests that this should work:
```python
with closing(Savepoint(cr)):
    ...
```
but the implementation does not match.  Can you make it work?

Reply:
The implementation should match, what doesn't work?

On exit, `closing` will call `Savepoint.close()` which will rollback the savepoint (unless it was already closed and thus released).

Reply:
The savepoint is rolled back, but not released.  That was my point.

Reply:
But it is? If the savepoint is not closed, `close` will conditionally rollback it, then unconditionally release it, then mark it as closed.

Reply:
Ah, after more testing turns out none of this works because `contextlib.closing` does not forward the entry, so in reality it never even creates the savepoint, and blows up on exit. Since while not the entire point `closing` is a pretty significant use case for the feature (I think) I guess that means you're getting your wish of the SAVEPOINT being issued when the object is initialised.

Means I need a `__del__` for cleanup, though I guess that can just release the savepoint (without rolling back).